include/revng/PTML/ContextualCommentMap.h

```diff
@@ -0,0 +1,223 @@
+/// get a (possibly empty) list of comments.
+///
+/// Don't forget to also call \ref getHomelessComments to get a (possibly empty)
+/// list containing all the comments that don't match a single node.
+template<HasStatementTraits NodeType>
+class ContextualCommentMap {
+private:
+  using Trait = StatementTraits<NodeType>;
+
+  using CommentType = model::ContextualComment;
+  using CommentLocationType = decltype(std::declval<CommentType>().Location());
+
+  struct NumericScore {
+    uint64_t Numerator = 0;
+    uint64_t Denominator = 1;
+
+    auto operator<=>(const NumericScore &Another) const {
+      uint64_t AdjustedLHS = this->Numerator * Another.Denominator;
+      uint64_t AdjustedRHS = Another.Numerator * this->Denominator;
+      return AdjustedLHS <=> AdjustedRHS;
+    }
+  };
+
+public:
+  struct CommentView {
+    uint64_t CommentIndex;
+    bool LocationMatchesExactly;
+    const std::decay_t<CommentLocationType> *ExpectedLocation;
+  };
+  using CommentList = std::vector<CommentView>;
+
+private:
+  // This assumes that `Trait::StatementType` is cheap (and, more importantly,
+  // safe) to copy
+  std::unordered_map<typename Trait::StatementType, CommentList> ResultMap;
+  CommentList HomelessComments;
+
+public:
+  ContextualCommentMap() = default;
+  ContextualCommentMap(const model::Function &Function, NodeType EntryNode) {
+    using LocationType = std::decay_t<typename Trait::LocationType>;
+    std::vector<std::pair<NumericScore, LocationType>> Scores;
+    Scores.resize(Function.Comments().size());
+    if (Scores.empty()) {
+      // No comments in this function, nothing to map.
+      return;
+    }
+
+    // First build a score map - select the best location for each comment
+    for (NodeType Node : llvm::depth_first(EntryNode)) {
+      for (const auto &Statement : Trait::getStatements(Node)) {
+        LocationType Location = Trait::getAddresses(Statement);
+
+        for (auto [Index, Comment] : llvm::enumerate(Function.Comments())) {
+
+          // Use the Tversky Index as the scoring function.
+          // Note the parameters controlling the weights of the sets: the higher
+          // a value is, the bigger the relative negative impact of
+          // the corresponding set having _extra_ elements compared to the other
+          // one.
+          //
+          // We want this because a line having unaccounted for addresses is a
+          // way bigger red flag that a (potentially user edited) comment having
+          // them.
+          auto Score = tverskyIndex<1, 2>(Comment.Location(), Location);
+
+          if (Score.has_value() && *Score > Scores[Index].first)
+            Scores[Index] = { *Score, Location };
+        }
+      }
+    }
+
+    // Then translate it into the mode-to-comment-list map taking into account
+    // the possibility of multiple nodes having the same location.
+    for (auto [Index, Comment] : llvm::enumerate(Function.Comments())) {
```

Comment:
`[]` thing is special in a sense that it is a universal reference by definition (`auto [X, Y]` is indistinguishable from `auto &&[X, Y]`). As such, it makes it so `Index` is copied (since it's an integer) and `Comment` is a reference, since that's what iterator's `operator*` returns.

Reply:
So `auto []` unpacking never creates a copy?

Reply:
Nope, the same way `auto &&` arguments never do. That's why it's the preferred way to accept "everything" in a heavily templated function.

Reply:
Ok, on the second thought, there are some special cases, think copiable-but-not-movable type returned by value. But, as you can probably imagine, those are insanely rare!

Reply:
Okay, it looks like I was wrong! Sorry about misinformation and good catch!

https://godbolt.org/z/771e1Ejhs

Reply:
My rule was "in `for` loops, use `const auto &[...]` if you don't want to change it, use `auto &[...]` if you want to change it, never use `auto [...]`".

Can I stick with it?

Reply:
I adopted blanket approach and fixed it everywhere formatting wasn't an issue. See the last commit I just pushed.